core/node/consensus/src/storage/registry_contract/committee_extractor.rs

```diff
@@ -0,0 +1,136 @@
+use anyhow::Context;
+use zksync_concurrency::{ctx, ctx::Ctx, scope, time};
+use zksync_consensus_roles::{attester, validator};
+use zksync_dal::consensus_dal;
+use zksync_node_api_server::execution_sandbox::BlockArgs;
+
+use crate::storage::{
+    registry_contract::{abi::CommitteeAttester, vm_reader::VMReader},
+    ConnectionPool,
+};
+
+#[derive(Debug)]
+pub struct CommitteeExtractor {
+    pool: ConnectionPool,
+    reader: VMReader,
+}
+
+impl CommitteeExtractor {
+    pub fn new(pool: ConnectionPool, reader: VMReader) -> CommitteeExtractor {
+        Self { pool, reader }
+    }
+
+    pub async fn run(mut self, ctx: &Ctx) -> anyhow::Result<()> {
+        tracing::info!("Running zksync_node_consensus::storage::CommitteeExtractor");
+        let res = scope::run!(ctx, |ctx, s| async {
+            let mut next_batch = self.next_batch_to_extract_committee(ctx).await?;
+            loop {
+                self.wait_batch(ctx, next_batch).await?;
+                self.extract_batch_committee(ctx, next_batch).await?;
+                next_batch = next_batch.next();
+            }
+        })
+        .await;
+        match res {
+            Ok(()) | Err(ctx::Error::Canceled(_)) => Ok(()),
+            Err(ctx::Error::Internal(err)) => Err(err),
+        }
+    }
+
+    async fn batch_max_l2_block_args(
+        &mut self,
+        ctx: &Ctx,
+        batch_number: attester::BatchNumber,
+    ) -> anyhow::Result<BlockArgs> {
+        let max_l2_block = self
+            .get_max_l2_block_of_l1_batch(ctx, batch_number)
+            .await?
+            .unwrap();
+
+        let block_number =
+            zksync_types::api::BlockNumber::Number(zksync_types::U64::from(max_l2_block.0));
+        let block_id = zksync_types::api::BlockId::Number(block_number);
+
+        self.reader
+            .block_args(ctx, block_id)
+            .await
+            .context("block_args()")
+    }
+
+    async fn extract_batch_committee(
+        &mut self,
+        ctx: &Ctx,
+        batch_number: attester::BatchNumber,
+    ) -> ctx::Result<()> {
+        let block_args = self.batch_max_l2_block_args(ctx, batch_number).await?;
+        let contract_attester_committee = match self.reader.contract_deployed(block_args).await {
+            false => vec![],
+            true => self
+                .reader
+                .read_attester_committee(block_args)
+                .await
+                .context("read_attester_committee()")?,
+        };
+
+        let db_attester_committee = consensus_dal::AttesterCommittee {
+            members: contract_attester_committee
+                .into_iter()
+                .map(CommitteeAttester::try_into)
+                .collect::<Result<Vec<consensus_dal::Attester>, _>>()?,
+        };
+
+        self.pool
+            .connection(ctx)
+            .await?
+            .insert_batch_committee(ctx, batch_number, db_attester_committee)
```

Comment:
Just checking my understanding of which batch number we mean here.

I thought the `committee` in `l1_batches_consensus` for batch number N means which is the committee that is expected to sign batch N. This is the result of executing batch N-1. 

So we call `extract_batch_committee` with the `next_batch` that we don't haven't extracted yet, and then figure out what is the last L2 block included in this batch. We execute the `eth_call` there to get the committee after the execution of the last L2 block. Then we insert the committee _with the same batch number_? Shouldn't it be `batch_number.inc()`?